includes/wc-cart-functions.php

```diff
@@ -31,30 +31,117 @@ function wc_protected_product_add_to_cart( $passed, $product_id ) {
+ * Add a product to the cart.
+ *
+ * Validates the product can be added, then adds it if valid.
+ *
+ * @since 2.7.0
+ * @param int $product_or_variation_id contains the id of the product to add to the cart.
+ * @param int $quantity contains the quantity of the item to add
+ * @param array $data extra cart item data we want to pass into the item, including any data about the variation if applicable.
+ * @return string|bool $cart_item_key that was added or false if validation failed.
  */
-function wc_empty_cart() {
-	if ( ! isset( WC()->cart ) || '' === WC()->cart ) {
-		WC()->cart = new WC_Cart();
+function wc_add_to_cart( $product_or_variation_id = 0, $quantity = 1, $data = array() ) {
+	try {
+		$product = wc_get_product( $product_or_variation_id );
+
+		if ( empty( $product ) || ! $product->is_purchasable() ) {
+			throw new Exception( __( 'Sorry, this product cannot be purchased.', 'woocommerce' ) );
+		}
+
+		$quantity = max( 0, apply_filters( 'woocommerce_add_to_cart_quantity', $quantity, $product, $data ) );
+		$data     = (array) apply_filters( 'woocommerce_add_to_cart_data', $data, $product );
+
+		wc_add_to_cart_validate_stock( $product, $quantity );
+
+		$cart_item_key = WC()->cart->add_item( array(
+			'product'  => $product,
+			'quantity' => $quantity,
+			'data'     => $data,
+		) );
+
+		do_action(
+			'woocommerce_add_to_cart',
+			$cart_item_key,
+			$product->get_id(),
+			$quantity,
+			is_callable( array( $product, 'get_variation_id' ) ) ? $product->get_variation_id() : 0,
+			! empty( $data['variation'] ) ? $data['variation'] : array(),
+			$data
+		);
+
+		return $cart_item_key;
+	} catch ( Exception $e ) {
+		wc_add_notice( $e->getMessage(), 'error' );
+		return false;
 	}
-	WC()->cart->empty_cart( false );
 }
 
 /**
- * Load the persistent cart.
- *
- * @param string $user_login
- * @param WP_User $user
- * @deprecated 2.3
+ * Check product has enough stock to be added to cart.
+ * @since 2.7.0
+ * @param WC_Product $product
+ * @param int $adding_quantity
+ * @throws Exception
  */
-function wc_load_persistent_cart( $user_login, $user ) {
-	if ( ! $user || ! ( $saved_cart = get_user_meta( $user->ID, '_woocommerce_persistent_cart', true ) ) ) {
-		return;
+function wc_add_to_cart_validate_stock( $product, $adding_quantity = 0 ) {
+	$products_qty_in_cart  = wc_cart_item_quantities();
+	$managing_stock        = $product->managing_stock();
+	$check_variation_stock = $product->is_type( 'variation' ) && true === $managing_stock;
+	$check_id              = $check_variation_stock ? $product->variation_id : $product->id;
+	$in_cart_qty           = isset( $products_qty_in_cart[ $check_id ] ) ? $products_qty_in_cart[ $check_id ] : 0;
+
+	if ( ! $product->is_in_stock() ) {
+		throw new Exception( sprintf( __( 'You cannot add &quot;%s&quot; to the cart because the product is out of stock.', 'woocommerce' ), $product->get_title() ) );
+	}
+
+	if ( $product->is_sold_individually() && $in_cart_qty && $adding_quantity ) {
```

Comment:
We don't really have a check for if there is no stock in qty, but adding_quantity is > 1.

This `if` check fails for that.

`if ( $product->is_sold_individually() && ( $in_cart_qty && $adding_quantity ) || ! $in_cart_qty && $adding_quantity > 1 ) {` works for both cases. or it could be a separate statement. See the failing `WC_Tests_Cart::test_add_to_cart_sold_individually` test about this one. (It adds 2 to the cart).